Services/communication_service.c

```diff
@@ -48,18 +60,156 @@ SAT_returnState communication_service_app(csp_packet_t *packet) {
       status = 0;
       memcpy(&packet->data[STATUS_BYTE], &status, sizeof(int8_t)); // 0 for success
       memcpy(&packet->data[OUT_DATA_BYTE], &temp_temp.sensor_temperature, sizeof(uint32_t));
-
+      // Step 4: set packet length
       set_packet_length(packet, sizeof(int8_t) + sizeof(uint32_t) + 1); // +1 for subservice
+      // Step 5: return packet
+      if (queue_response(packet) != SATR_OK) {
+        return SATR_ERROR;
+      }
+
+      break;
 
+    case GET_FREQ:
+      HAL_S_getFreq(&S_config.S_freq);
+      S_config.S_freq = csp_hton32(S_config.S_freq);
+      status = 0;
+      memcpy(&packet->data[STATUS_BYTE], &status, sizeof(int8_t));
+      memcpy(&packet->data[OUT_DATA_BYTE], &S_config.S_freq, sizeof(uint32_t));
+      set_packet_length(packet, sizeof(int8_t) + sizeof(uint32_t) + 1);
       if (queue_response(packet) != SATR_OK) {
         return SATR_ERROR;
       }
+      break;
 
+    case GET_CONTROL:
+      HAL_S_getControl (&S_PA.PA_status, &S_PA.PA_mode);
+      S_PA.PA_status = csp_hton32(S_PA.PA_status);
+      S_PA.PA_mode = csp_hton32(S_PA.PA_mode);
+      status = 0;
+      memcpy(&packet->data[STATUS_BYTE], &status, sizeof(int8_t));
+      memcpy(&packet->data[OUT_DATA_BYTE], &S_PA.PA_status, sizeof(uint32_t));
+      memcpy(&packet->data[OUT_DATA_BYTE + sizeof(uint32_t)], &S_PA.PA_mode, sizeof(uint32_t));
+      set_packet_length(packet, sizeof(int8_t) + sizeof(uint32_t) + sizeof(uint32_t) + 1);
+
+      if (queue_response(packet) != SATR_OK) {
+        return SATR_ERROR;
+      }
       break;
 
+    case GET_ENCODER:
+      HAL_S_getEncoder(&S_enc.Enc_scrambler, &S_enc.Enc_filter, &S_enc.Enc_modulation, &S_enc.Enc_rate);
+      S_enc.Enc_scrambler = csp_hton32(S_enc.Enc_scrambler);
+      S_enc.Enc_filter = csp_hton32(S_enc.Enc_filter);
+      S_enc.Enc_modulation = csp_hton32(S_enc.Enc_modulation);
+      S_enc.Enc_rate = csp_hton32(S_enc.Enc_rate);
+      status = 0;
+
+      memcpy(&packet->data[STATUS_BYTE], &status, sizeof(int8_t));
+      memcpy(&packet->data[OUT_DATA_BYTE], &S_enc.Enc_scrambler, sizeof(int8_t));
+      memcpy(&packet->data[OUT_DATA_BYTE + sizeof(int8_t)], &S_enc.Enc_filter, sizeof(int8_t));
+      memcpy(&packet->data[OUT_DATA_BYTE + 2*sizeof(int8_t)], &S_enc.Enc_modulation, sizeof(int8_t));
+      memcpy(&packet->data[OUT_DATA_BYTE + 3*sizeof(int8_t)], &S_enc.Enc_rate, sizeof(int8_t));
```

Comment:
Load the entire struct at once

Reply:
For copying the data we can but for csp_hton functions, they don't accept structures. Would it be ok to convert them individually?

Reply:
Yeah I don't think there's a better way